src/GafferScene/RenderController.cpp

```diff
@@ -958,9 +958,15 @@ class RenderController::SceneGraph
 			m_objectHash = MurmurHash();
 		}
 
-		bool updateExpansion( const ScenePlug::ScenePath &path, const IECore::PathMatcher &expandedPaths, size_t minimumExpansionDepth )
+		bool updateExpansion( const ScenePlug::ScenePath &path, const GafferScene::VisibleSet &visibleSet, size_t minimumExpansionDepth )
 		{
-			const bool expanded = ( minimumExpansionDepth >= path.size() ) || ( expandedPaths.match( path ) & PathMatcher::ExactMatch );
+			const bool expanded =
+				( minimumExpansionDepth >= path.size() && !( visibleSet.exclusions.match( path ) & ( PathMatcher::ExactMatch | PathMatcher::AncestorMatch ) ) ) ||
+				/// \todo Using EveryMatch ensures the ancestors of paths in `visibleSet.inclusions` are also matched for expansion
+				/// ideally we'd avoid expanding those ancestors and instead render only the included path and its descendants.
+				visibleSet.match( path ) & PathMatcher::EveryMatch
+			;
```

Comment:
How about we add a `minimumExpansionDepth = 0` argument to `VisibleSet::match()`, so that we get to keep the visibility logic all in one place?

Reply:
Nice idea! I've tried that in 77efe09 and it definitely simplifies things for the RenderController.